corehq/motech/auth.py

```diff
@@ -0,0 +1,38 @@
+import requests
+import re
+from urllib.parse import urljoin
+from requests.exceptions import RequestException
+
+
+class HTTPBearerAuth(requests.auth.AuthBase):
+    def __init__(self, username, plaintext_password):
+        self.username = username
+        self.password = plaintext_password
+
+    def _find_bearer_base(self, r):
+        m = re.compile('https.*/api/v[0-9]+/').match(r.url)
+        if m:
+            return m.group(0)
+        else:
+            raise RequestException(None, r, "HTTP endpoint is not not valid for bearer auth")
+
+    def _get_auth_token(self, r):
+        token_base = self._find_bearer_base(r)
+        token_request_url = urljoin(token_base, "token")
+
+        post_data = {"grant_type": "password",
+            "username": self.username,
+            "password": self.password
+                    }
+
+        token_response = requests.get(token_request_url, data=post_data)
+        try:
+            return token_response.json()['access_token']
+        except Exception:
+            raise RequestException(None, r,
+                    "Unable to retrieve access token for request: %s" % token_response.content)
+
+    def __call__(self, r):
+        token = self._get_auth_token(r)
+        r.headers["Authorization"] = "Bearer %s" % token
```

Comment:
I agree that this isn't great. I think this could be done by storing the token and related details in the ConnectionSettings / Repeater (perhaps as a json field or something). Then we can pass the ConnectionSettings to `HTTPBearerAuth`. If the token doesn't exist or has expired we get a new one and save it into the ConnectionSettings.

Reply:
I also think the `Requests` class might try to use basic auth regardless and/or mess with this. I'll check this branch out, and see what headers the requests instance passes. Do we have an endpoint we can test with @ctsims ? If not, I'll see if I can find one.

Reply:
@kaapstorm  - This does work, and I've tested it against the external API, so no worries on BasicAuth overriding anything. I can share the endpoint/credentials but we should be a bit cautious, it got pretty annoyed with me when I was testing error handling and rate limited me out pretty fast when I auth'd badly.

Reply:
@kaapstorm Agreed it would be good to hold onto the token if possible, they do also issue refresh tokens through the API. 

Currently the tokens only last ~15 minutes, and our requests will be 30 minutes apart, though, so not sure if urgent.

Reply:
this is what I was thinking of (bear bones): https://github.com/dimagi/commcare-hq/pull/27263